src/client/id_detection.go

```diff
@@ -8,18 +8,62 @@ import (
 )
 
 const (
-	consumerDetectionPattern = "kafka.consumer:type=consumer-fetch-manager-metrics,client-id=*"
-	producerDetectionPattern = "kafka.producer:type=producer-metrics,client-id=*"
+	consumerAppInfoPattern = "kafka.consumer:type=app-info,id=*"
+	consumerMetricsPattern = "kafka.consumer:type=consumer-fetch-manager-metrics,client-id=*"
+
+	producerAppInfoPattern = "kafka.producer:type=app-info,id=*"
+	producerMetricsPattern = "kafka.producer:type=producer-metrics,client-id=*"
 )
 
 // idFromMBeanNameFn defines a function to extract the identifier from an MBean name.
 type idFromMBeanNameFn func(string) string
 
-func getClientIDS(jmxInfo *args.JMXHost, mBeanPattern string, idExtractor idFromMBeanNameFn, conn connection.JMXConnection) ([]string, error) {
+type clientIDExtractInfo struct {
+	pattern   string
+	extractor idFromMBeanNameFn
+}
+
+func detectConsumerIDs(jmxInfo *args.JMXHost, conn connection.JMXConnection) ([]string, error) {
+	// consumerAppInfoPatter is defined to detect consumer clientIDs, in case it is not found it will use
+	// consumerMetricsPattern which is also being used to fetch consumer metrics.
+	return getClientIDS(
+		[]clientIDExtractInfo{
+			{pattern: consumerAppInfoPattern, extractor: idFromAppInfo},
+			{pattern: consumerMetricsPattern, extractor: idFromMBeanWithClientIDField},
+		},
+		jmxInfo,
+		conn,
+	)
+}
+
+func detectProducerIDs(jmxInfo *args.JMXHost, conn connection.JMXConnection) ([]string, error) {
+	// producerAppInfoPatter is defined to detect consumer clientIDs, in case it is not found it will use
+	// ProducerMetricsPattern which is also being used to fetch consumer metrics.
+	return getClientIDS(
+		[]clientIDExtractInfo{
+			{pattern: producerAppInfoPattern, extractor: idFromAppInfo},
```

Comment:
Is the "app-info" more correct than the other one used before? Shouldn't we modify the order and try first by the other and have the app-info as the fallback? In the official documentation, they talk about the other name https://docs.confluent.io/3.2.2/kafka/monitoring.html#fetch-metrics.

Reply:
I though the app-info was some kind of java/jmx standard, but as long as we are fetching metrics I think you are right. Maybe we could even try the other patterns we are currently using to fetch metrics before the app-info, in: [producers_definitions](https://github.com/newrelic/nri-kafka/blob/edd6f0986c458612b9acd7e71c8f782dbb9936f1/src/metrics/producer_definitions.go#L1) and [consumers_definition](https://github.com/newrelic/nri-kafka/blob/edd6f0986c458612b9acd7e71c8f782dbb9936f1/src/metrics/consumer_definitions.go).

WDYT?